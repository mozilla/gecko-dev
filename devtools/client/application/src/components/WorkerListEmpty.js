/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { openDocLink, openTrustedLink } = require("devtools/client/shared/link");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const { createFactory, Component } = require("devtools/client/shared/vendor/react");
const { a, article, h1, li, p, ul } = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const DOC_URL = "https://developer.mozilla.org/docs/Web/API/Service_Worker_API/Using_Service_Workers" +
  "?utm_source=devtools&utm_medium=sw-panel-blank";

/**
 * This component displays help information when no service workers are found for the
 * current target.
 */
class WorkerListEmpty extends Component {
  static get propTypes() {
    return {
      serviceContainer: PropTypes.object.isRequired,
    };
  }

  switchToConsole() {
    this.props.serviceContainer.selectTool("webconsole");
  }

  switchToDebugger() {
    this.props.serviceContainer.selectTool("jsdebugger");
  }

  openAboutDebugging() {
    openTrustedLink("about:debugging#workers");
  }

  openDocumentation() {
    openDocLink(DOC_URL);
  }

  render() {
    return article(
      { className: "worker-list-empty" },
      Localized({
        id: "serviceworker-empty-intro",
        a: a({ className: "external-link", onClick: () => this.openDocumentation() }),
      },
        h1({ className: "worker-list-empty__title" })
      ),
      Localized(
        { id: "serviceworker-empty-suggestions" },
        p({})
      ),
      ul(
        { className: "worker-list-empty__tips" },
        Localized({
          id: "serviceworker-empty-suggestions-console",
          a: a({ className: "link", onClick: () => this.switchToConsole() }),
        },
          li({ className: "worker-list-empty__tips__item" })
        ),
        Localized({
          id: "serviceworker-empty-suggestions-debugger",
          a: a({ className: "link", onClick: () => this.switchToDebugger() }),
        },
          li({ className: "worker-list-empty__tips__item" })
        ),
        Localized({
          id: "serviceworker-empty-suggestions-aboutdebugging",
          a: a({
            className: "link js-trusted-link",
            onClick: () => this.openAboutDebugging() }),
        },
          li({ className: "worker-list-empty__tips__item" })
        ),
      )
    );
  }
}

// Exports

module.exports = WorkerListEmpty;
