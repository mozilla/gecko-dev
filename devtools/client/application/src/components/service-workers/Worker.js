/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

const {
  br,
  dd,
  dl,
  dt,
  header,
  li,
  section,
  span,
  time,
} = require("devtools/client/shared/vendor/react-dom-factories");

const {
  getUnicodeUrl,
  getUnicodeUrlPath,
} = require("devtools/client/shared/unicode-url");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const Types = require("../../types/index");

const UIButton = createFactory(require("../ui/UIButton"));

loader.lazyRequireGetter(
  this,
  "gDevToolsBrowser",
  "devtools/client/framework/devtools-browser",
  true
);

/**
 * This component is dedicated to display a worker, more accurately a service worker, in
 * the list of workers displayed in the application panel. It displays information about
 * the worker as well as action links and buttons to interact with the worker (e.g. debug,
 * unregister, update etc...).
 */
class Worker extends PureComponent {
  static get propTypes() {
    return {
      isDebugEnabled: PropTypes.bool.isRequired,
      worker: PropTypes.shape(Types.worker).isRequired,
    };
  }

  constructor(props) {
    super(props);

    this.debug = this.debug.bind(this);
    this.start = this.start.bind(this);
    this.unregister = this.unregister.bind(this);
  }

  debug() {
    if (!this.isRunning()) {
      console.log("Service workers cannot be debugged if they are not running");
      return;
    }

    const { workerTargetFront } = this.props.worker;
    gDevToolsBrowser.openWorkerToolbox(workerTargetFront);
  }

  start() {
    if (!this.props.isDebugEnabled) {
      console.log("Service workers cannot be started in multi-e10s");
      return;
    }

    if (!this.isActive() || this.isRunning()) {
      console.log("Running or inactive service workers cannot be started");
      return;
    }

    const { registrationFront } = this.props.worker;
    registrationFront.start();
  }

  unregister() {
    const { registrationFront } = this.props.worker;
    registrationFront.unregister();
  }

  isRunning() {
    // We know the worker is running if it has a worker actor.
    return !!this.props.worker.workerTargetFront;
  }

  isActive() {
    return this.props.worker.active;
  }

  getServiceWorkerStatus() {
    if (this.isActive() && this.isRunning()) {
      return "running";
    } else if (this.isActive()) {
      return "stopped";
    }
    // We cannot get service worker registrations unless the registration is in
    // ACTIVE state. Unable to know the actual state ("installing", "waiting"), we
    // display a custom state "registering" for now. See Bug 1153292.
    return "registering";
  }

  formatScope(scope) {
    const [, remainder] = getUnicodeUrl(scope).split("://");
    return remainder || scope;
  }

  formatSource(source) {
    const parts = source.split("/");
    return getUnicodeUrlPath(parts[parts.length - 1]);
  }

  renderDebugButton() {
    const { isDebugEnabled } = this.props;

    const isDisabled = !this.isRunning() || !isDebugEnabled;

    const localizationId = isDebugEnabled
      ? "serviceworker-worker-debug"
      : "serviceworker-worker-debug-forbidden";

    return Localized(
      {
        id: localizationId,
        // The localized title is only displayed if the debug link is disabled.
        attrs: {
          title: isDisabled,
        },
      },
      UIButton({
        onClick: this.debug,
        className: `js-debug-button`,
        disabled: isDisabled,
        size: "micro",
      })
    );
  }

  renderStartButton() {
    const { isDebugEnabled } = this.props;
    const isDisabled = !isDebugEnabled;

    return Localized(
      {
        id: "serviceworker-worker-start2",
        // The localized title is only displayed if the debug link is disabled.
        attrs: {
          title: !isDisabled,
        },
      },
      UIButton({
        onClick: this.start,
        className: `js-start-button`,
        disabled: isDisabled,
        size: "micro",
      })
    );
  }

  render() {
    const { worker } = this.props;
    const status = this.getServiceWorkerStatus();

    const unregisterButton = this.isActive()
      ? Localized(
          { id: "serviceworker-worker-unregister" },
          UIButton({
            onClick: this.unregister,
            className: "worker__unregister-button js-unregister-button",
          })
        )
      : null;

    const lastUpdated = worker.lastUpdateTime
      ? Localized(
          {
            id: "serviceworker-worker-updated",
            // XXX: $date should normally be a Date object, but we pass the timestamp as a
            // workaround. See Bug 1465718. worker.lastUpdateTime is in microseconds,
            // convert to a valid timestamp in milliseconds by dividing by 1000.
            $date: worker.lastUpdateTime / 1000,
            time: time({ className: "js-sw-updated" }),
          },
          span({ className: "worker__data__updated" })
        )
      : null;

    return li(
      { className: "worker js-sw-container" },
      header(
        { className: "worker__header" },
        span(
          { title: worker.scope, className: "worker__scope js-sw-scope" },
          this.formatScope(worker.scope)
        ),
        section({ className: "worker__controls" }, unregisterButton)
      ),
      dl(
        { className: "worker__data" },
        Localized(
          { id: "serviceworker-worker-source" },
          dt({ className: "worker__meta-name" })
        ),
        dd(
          {},
          span(
            {
              title: worker.scope,
              className: "worker__source-url js-source-url",
            },
            this.formatSource(worker.url)
          ),
          " ",
          this.renderDebugButton(),
          lastUpdated ? br({}) : null,
          lastUpdated ? lastUpdated : null
        ),
        Localized(
          { id: "serviceworker-worker-status" },
          dt({ className: "worker__meta-name" })
        ),
        dd(
          {},
          Localized(
            { id: "serviceworker-worker-status-" + status },
            span({ className: "js-worker-status" })
          ),
          " ",
          !this.isRunning() ? this.renderStartButton() : null
        )
      )
    );
  }
}

module.exports = Worker;
