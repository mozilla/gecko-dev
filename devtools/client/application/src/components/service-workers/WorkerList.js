/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { openTrustedLink } = require("devtools/client/shared/link");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  a,
  article,
  footer,
  h1,
  ul,
} = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const Types = require("../../types/index");
const Worker = createFactory(require("./Worker"));

/**
 * This component handles the list of service workers displayed in the application panel
 * and also displays a suggestion to use about debugging for debugging other service
 * workers.
 */
class WorkerList extends PureComponent {
  static get propTypes() {
    return {
      canDebugWorkers: PropTypes.bool.isRequired,
      workers: Types.workerArray.isRequired,
    };
  }

  render() {
    const { canDebugWorkers, workers } = this.props;

    return [
      article(
        { className: "workers-container", key: "workers-container" },
        Localized(
          { id: "serviceworker-list-header" },
          h1({
            className: "app-page__title",
          })
        ),
        ul(
          {},
          workers.map(worker =>
            Worker({
              key: worker.id,
              isDebugEnabled: canDebugWorkers,
              worker,
            })
          )
        )
      ),
      Localized(
        {
          id: "serviceworker-list-aboutdebugging",
          key: "serviceworkerlist-footer",
          a: a({
            className: "aboutdebugging-plug__link",
            onClick: () => openTrustedLink("about:debugging#workers"),
          }),
        },
        footer({ className: "aboutdebugging-plug" })
      ),
    ];
  }
}

// Exports
module.exports = WorkerList;
