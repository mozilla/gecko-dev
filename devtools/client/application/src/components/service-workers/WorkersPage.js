/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  section,
} = require("devtools/client/shared/vendor/react-dom-factories");
const { connect } = require("devtools/client/shared/vendor/react-redux");

const Types = require("../../types/index");
const WorkerList = createFactory(require("./WorkerList"));
const WorkerListEmpty = createFactory(require("./WorkerListEmpty"));

class WorkersPage extends PureComponent {
  static get propTypes() {
    return {
      // mapped from state
      canDebugWorkers: PropTypes.bool.isRequired,
      domain: PropTypes.string.isRequired,
      workers: Types.workerArray.isRequired,
    };
  }

  render() {
    const { canDebugWorkers, domain, workers } = this.props;

    // Filter out workers from other domains
    const domainWorkers = workers.filter(
      x => new URL(x.url).hostname === domain
    );
    const isWorkerListEmpty = domainWorkers.length === 0;

    return section(
      {
        className: `app-page js-service-workers-page ${
          isWorkerListEmpty ? "app-page--empty" : ""
        }`,
      },
      isWorkerListEmpty
        ? WorkerListEmpty({})
        : WorkerList({
            canDebugWorkers,
            workers: domainWorkers,
          })
    );
  }
}

function mapStateToProps(state) {
  return {
    canDebugWorkers: state.workers.canDebugWorkers,
    domain: state.page.domain,
    workers: state.workers.list,
  };
}

// Exports
module.exports = connect(mapStateToProps)(WorkersPage);
