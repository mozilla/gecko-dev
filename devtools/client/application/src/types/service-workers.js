/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

const worker = {
  active: PropTypes.bool,
  name: PropTypes.string.isRequired,
  scope: PropTypes.string.isRequired,
  lastUpdateTime: PropTypes.number.isRequired,
  url: PropTypes.string.isRequired,
  // registrationFront can be missing in e10s.
  registrationFront: PropTypes.object,
  workerTargetFront: PropTypes.object,
};

const workerArray = PropTypes.arrayOf(PropTypes.shape(worker));

module.exports = {
  worker,
  workerArray,
};
