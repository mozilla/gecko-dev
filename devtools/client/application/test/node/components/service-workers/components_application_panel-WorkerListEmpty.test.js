/* Any copyright is dedicated to the Public Domain.
   http://creativecommons.org/publicdomain/zero/1.0/ */

"use strict";

// Import libs
const { shallow } = require("enzyme");
const { createFactory } = require("react");

const WorkerListEmpty = createFactory(
  require("devtools/client/application/src/components/service-workers/WorkerListEmpty")
);

/**
 * Test for workerListEmpty.js component
 */

describe("WorkerListEmpty", () => {
  it("renders the expected snapshot", () => {
    const wrapper = shallow(WorkerListEmpty({}));
    expect(wrapper).toMatchSnapshot();
  });
});
