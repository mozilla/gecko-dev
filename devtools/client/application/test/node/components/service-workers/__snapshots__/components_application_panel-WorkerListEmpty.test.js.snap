// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkerListEmpty renders the expected snapshot 1`] = `
<article
  className="worker-list-empty js-worker-list-empty"
>
  <Localized
    a={
      <a
        className="external-link"
        onClick={[Function]}
      />
    }
    id="serviceworker-empty-intro"
  >
    <h1
      className="app-page__title"
    />
  </Localized>
  <Localized
    id="serviceworker-empty-suggestions"
  >
    <p />
  </Localized>
  <ul
    className="worker-list-empty__tips"
  >
    <Localized
      a={
        <a
          className="link"
          onClick={[Function]}
        />
      }
      id="serviceworker-empty-suggestions-console"
    >
      <li
        className="worker-list-empty__tips__item"
      />
    </Localized>
    <Localized
      a={
        <a
          className="link"
          onClick={[Function]}
        />
      }
      id="serviceworker-empty-suggestions-debugger"
    >
      <li
        className="worker-list-empty__tips__item"
      />
    </Localized>
    <Localized
      a={
        <a
          className="link js-trusted-link"
          onClick={[Function]}
        />
      }
      id="serviceworker-empty-suggestions-aboutdebugging"
    >
      <li
        className="worker-list-empty__tips__item"
      />
    </Localized>
  </ul>
</article>
`;
